client/shared/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -289,16 +290,37 @@ object Client {
       go(source).stream
     }
 
-    def runOrDispose(req: Request[F]): F[Resource[F, Response[F]]] =
-      Ref[F].of(false).map { disposed =>
-        val req0 = addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
-        Resource
-          .eval(app(req0))
-          .onFinalize(disposed.set(true))
-          .map(_.pipeBodyThrough(until(disposed)))
-      }
+    def run(
+        message: Message[F],
+        refOp: Option[Ref[F, Boolean]] = None,
+    ): Resource[F, Response[F]] = message match {
+      case req: Request[F] =>
+        Resource.eval(Ref[F].of(false)).flatMap { disposed =>
+          val reqAugmented =
+            addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
+          Resource
+            .eval(app(reqAugmented))
+            .onFinalize(disposed.set(true))
+            .flatMap(run(_, Option(disposed)))
+        }
+      case resp: Response[F] =>
+        Resource.eval(refOp.fold(Ref[F].of(false))(F.pure)).flatMap { r =>
+          Resource
+            .pure(resp.pipeBodyThrough(until(r)(_).onFinalizeCase {
+              case ExitCase.Canceled | ExitCase.Errored(_) => F.unit
+              case ExitCase.Succeeded => r.set(true)
+            }))
```

Comment:
I need to think about this some more. What happens if there is an error / cancellation while the client is consuming the `Stream`? In that case, it will compile it again in the `onFinalize` before. But what bothers me is then it may be starting the stream again from the beginning.

I wonder if what we need to do is create an intermediate `fs2.concurrent.Channel` to drain the `Stream` into. That way, even if the client errors/cancels half-way through, the final drain at the end will pick up from where it left off, instead of recompiling from scratch.

It would be really great to add a test case to cover this situation.

Reply:
Oh! I think there may be another case. What happens if the stream ends successfully, without draining all the bytes? e.g. `stream.take(1).compile.drain` 🤔

I think a `Channel` might be the best solution here.

Reply:
Yep, I think you are right. I will try to solve it with `Channel`, thanks for the idea!